src/Http/Http.Abstractions/src/Internal/PathStringHelper.cs

```diff
@@ -1,47 +1,64 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
 namespace Microsoft.AspNetCore.Http.Internal
 {
-    internal class PathStringHelper
+    internal static class PathStringHelper
     {
-        private static bool[] ValidPathChars = {
-            false, false, false, false, false, false, false, false,     // 0x00 - 0x07
-            false, false, false, false, false, false, false, false,     // 0x08 - 0x0F
-            false, false, false, false, false, false, false, false,     // 0x10 - 0x17
-            false, false, false, false, false, false, false, false,     // 0x18 - 0x1F
-            false, true,  false, false, true,  false, true,  true,      // 0x20 - 0x27
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x28 - 0x2F
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x30 - 0x37
-            true,  true,  true,  true,  false, true,  false, false,     // 0x38 - 0x3F
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x40 - 0x47
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x48 - 0x4F
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x50 - 0x57
-            true,  true,  true,  false, false, false, false, true,      // 0x58 - 0x5F
-            false, true,  true,  true,  true,  true,  true,  true,      // 0x60 - 0x67
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x68 - 0x6F
-            true,  true,  true,  true,  true,  true,  true,  true,      // 0x70 - 0x77
-            true,  true,  true,  false, false, false, true,  false,     // 0x78 - 0x7F
+        private static readonly uint[] ValidPathChars = {
+            0b_0000_0000__0000_0000__0000_0000__0000_0000, // 0x00 - 0x1F
+            0b_0010_1111__1111_1111__1111_1111__1101_0010, // 0x20 - 0x3F
+            0b_1000_0111__1111_1111__1111_1111__1111_1111, // 0x40 - 0x5F
+            0b_0100_0111__1111_1111__1111_1111__1111_1110, // 0x60 - 0x7F
         };
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsValidPathChar(char c)
         {
-            return c < ValidPathChars.Length && ValidPathChars[c];
+            // Use local array and uint .Length compare to elide the bounds check on array access
+            var validChars = ValidPathChars;
+            var i = (int)c;
+
+            // Array is in chunks of 32 bits, so get offset by dividing by 32
+            var offset = i / 32;
```

Comment:
```suggestion
            var offset = i >> 5;		// i / 32;
```
AFAIK JIT does this optimization only for [unsigned](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwGZ8AmXAYVwG9zc39qBLAOzlwFkAFDz6IAlLgC8APlyJcAelyUapMuw64AriIGCdvORJlzFy1eQC+QA=)